### ISSUE TITLE
promotion dtype warnings when accessing a FITS header with a Variable-Length Field

### ISSUE THREAD

**Issue by DougBurke** [astropy/astropy#16819]
### Description

I am seeing `UserWarning` messages from AstroPy from NumPy 2.0 with the `NPY_PROMOTION_STATE=weak_and_warn` environment variable set. There appears to be an issue when reading in a Variable-Length column - which I unfortunately have to do because that's what X-ray RMF data often uses.

The warning messages are of the form `UserWarning: result dtype changed due to the removal of value-based promotion from NumPy. Changed from int64 to int32.`

Part of https://github.com/sherpa/sherpa/issues/2087

### Expected behavior

Not to get warning messages when run with the environment NPY_PROMOTION_STATE set to `weak_and_warn`. 

### How to Reproduce

With the following script

```python
import importlib.metadata
import os

from astropy.io import fits

print(f"## flag: NPY_PROMOTION_STATE='{os.getenv('NPY_PROMOTION_STATE')}'")

print(f"## Using AstroPy: {importlib.metadata.version('astropy')}")
print(f"##       NumPy:   {importlib.metadata.version('numpy')}")

# Open a file using variable-length columns, because why would using
# Astronomical data be easy....

hdus = fits.open("https://github.com/sherpa/sherpa-test-data/raw/main/sherpatest/3c273.rmf")
print("")

print("# Columns:")
matrix = hdus["MATRIX"]
print(matrix.columns)
print("")

# Read in the F_CHAN column
#
col = matrix.data["F_CHAN"]
print("")

# Display the first three rows:
print(col[:3])

print("")
print("# Fini")
```

I get the following output

```
## flag: NPY_PROMOTION_STATE='weak_and_warn'
## Using AstroPy: 6.1.2
##       NumPy:   2.0.1

# Columns:
ColDefs(
    name = 'ENERG_LO'; format = 'E'; unit = 'keV'
    name = 'ENERG_HI'; format = 'E'; unit = 'keV'
    name = 'N_GRP'; format = 'I'
    name = 'F_CHAN'; format = 'PI(2)'
    name = 'N_CHAN'; format = 'PI(2)'
    name = 'MATRIX'; format = 'PE(81)'
)

/home/dburke/micromamba/envs/sherpa-conda/lib/python3.11/site-packages/astropy/io/fits/fitsrec.py:806: UserWarning: result dtype changed due to the removal of value-based promotion from NumPy. Changed from int64 to int32.
  offset = field[idx, 1] + self._heapoffset
/home/dburke/micromamba/envs/sherpa-conda/lib/python3.11/site-packages/astropy/io/fits/fitsrec.py:817: UserWarning: result dtype changed due to the removal of value-based promotion from NumPy. Changed from int64 to int32.
  arr_len = count * dt.itemsize

[array([8], dtype=int16) array([8], dtype=int16) array([8], dtype=int16)]

# Fini
```

### Versions

```python
import astropy
try:
    astropy.system_info()
except AttributeError:
    import platform; print(platform.platform())
    import sys; print("Python", sys.version)
    import astropy; print("astropy", astropy.__version__)
    import numpy; print("Numpy", numpy.__version__)
    import erfa; print("pyerfa", erfa.__version__)
    try:
        import scipy
        print("Scipy", scipy.__version__)
    except ImportError:
        print("Scipy not installed")
    try:
        import matplotlib
        print("Matplotlib", matplotlib.__version__)
    except ImportError:
        print("Matplotlib not installed")
```
```
# Paste the result here
Linux-6.5.0-1027-oem-x86_64-with-glibc2.35
Python 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0]
astropy 6.1.2
Numpy 2.0.1
pyerfa 2.0.1.4
Scipy not installed
Matplotlib 3.9.0
```

**Comment by neutrinoceros** [comment]
Hi @DougBurke, thank you for reporting !

Quoting NEP 50:
>  Note that weak_and_warn implements the optional warnings proposed in this NEP and is expected to be very noisy. We recommend starting using the weak option and use weak_and_warn mainly to understand a specific observed change in behaviour.

It sounds like getting rid of all such warnings could be quite an undertaking.
Can you explain a bit more your motivation here ? Thanks !

**Comment by DougBurke** [comment]
So, the implicit conversion between data types can (I believe) be a subtle source of bugs when the rules change, as they have done here.

I have an old python code base that I am trying to make sure continues working correctly, so I am trying to be forward looking and using facilities such as `NPY_PROMOTION_STATE` to help me out. Unfortunately our test suite loads a load of RMF files and so any signal of a problem I am looking for in my code base is hard to find because there are all these warnings from the FITS code in astropy.

So, I don't have any sign that this change is causing any error. In an ideal world this warning would go away, but I understand that this is unlikely to be a high-priority issue.

**Comment by neutrinoceros** [comment]
The fix in the case you gave is actually simple enough that I don't see any reason not to go for it:
```patch
diff --git a/astropy/io/fits/fitsrec.py b/astropy/io/fits/fitsrec.py
index ae364f9414..9e56d5369d 100644
--- a/astropy/io/fits/fitsrec.py
+++ b/astropy/io/fits/fitsrec.py
@@ -803,18 +803,18 @@ class FITS_rec(np.recarray):
             )
 
         for idx in range(len(self)):
-            offset = field[idx, 1] + self._heapoffset
+            offset = field[idx, 1] + field.dtype.type(self._heapoffset)
             count = field[idx, 0]
 
             if recformat.dtype == "S":
                 dt = np.dtype(recformat.dtype + str(1))
-                arr_len = count * dt.itemsize
+                arr_len = count * count.dtype.type(dt.itemsize)
                 da = raw_data[offset : offset + arr_len].view(dt)
                 da = np.char.array(da.view(dtype=dt), itemsize=count)
                 dummy[idx] = decode_ascii(da)
             else:
                 dt = np.dtype(recformat.dtype)
-                arr_len = count * dt.itemsize
+                arr_len = count * count.dtype.type(dt.itemsize)
                 dummy[idx] = raw_data[offset : offset + arr_len].view(dt)
                 if column.dim and len(vla_shape) > 1:
                     # The VLA is reshaped consistently with TDIM instructions
```

what I'm worried about is whether this would be sufficient to help you or not.

**Comment by neutrinoceros** [comment]
The specific patch I posted shouldn't have any side effect: it's equivalent to letting numpy apply its type promotion rules itself. And if it works on numpy 2 there's no reason to think that it could fail on numpy 1.

Just for fun I tried
```
NPY_PROMOTION_STATE=weak_and_warn pytest -n5
```
locally and got 458 failures (down to 450 with my patch applied). We could in principle address all of them but I don't think it'd be worth the time and effort.
How about I open a PR with this patch and we run extra CI on it ? (provided @DougBurke is satisfied with us fixing *just* this bit).
